autorest/adal/token.go

```diff
@@ -534,8 +537,13 @@ func newTokenRefreshError(message string, resp *http.Response) TokenRefreshError
 // EnsureFresh will refresh the token if it will expire within the refresh window (as set by
 // RefreshWithin) and autoRefresh flag is on.
 func (spt *ServicePrincipalToken) EnsureFresh() error {
-	if spt.autoRefresh && spt.WillExpireIn(spt.refreshWithin) {
-		return spt.Refresh()
+	if spt.autoRefresh && spt.token.WillExpireIn(spt.refreshWithin) {
+		// take the write lock then check to see if the token was already refreshed
+		spt.refreshLock.Lock()
+		defer spt.refreshLock.Unlock()
+		if spt.token.WillExpireIn(spt.refreshWithin) {
```

Comment:
>	if spt.token.WillExpireIn(spt.refreshWithin) { [](start = 1, length = 47)

we do this extra check because the token probably it was already refreshed  if we got stock on the lock ?

Reply:
Since the write lock serializes the refresh, we check again once we hold it to see if some other goroutine already refreshed the token (no point in refreshing again).

Reply:
sounds good